### PR TITLE
chore: upgrade to polkadot 0.9.11

### PR DESCRIPTION
**Pull request by wischli**
## fixes KILTProtocol/ticket#1630
* Upgrades to Polkadot v0.9.11 which adds the `TypeInfo` to on chain data structs which automatically adds it to the metadata and removes the necessity of providing type data separately
* Applies Vesting refactor which changed the StorageType from being a single struct to a BoundedVec

Changes for `TypeInfo` and metadata where inspired by the corresponding [migration guide](https://gist.github.com/ascjones/0d81a4c44e84cacd9f714cd34a6de823).

**Note: We need to merge this ASAP and update our clients before [0.9.12](https://github.com/paritytech/polkadot/issues/4095) goes live on Kusama. Else we cannot sync anymore.**

## TODO

- [x] Unit tests
- [x] Bench pallet_vesting
- [ ] Bench pallet_democracy
- [x] Bench pallet_balances
- [x] KILT Launch pallet

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)
^

### CODE REVIEW

File: pallets/ctype/Cargo.toml [hunk 1]
```diff
@@ -14,36 +14,38 @@ substrate-wasm-builder-runner = {version = "3.0.0"}
...
-frame-support = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate"}
-frame-system = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate"}
-pallet-balances = {optional = true, branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate"}
-sp-core = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate", optional = true}
-sp-io = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate", optional = true}
-sp-keystore = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate", optional = true}
-sp-runtime = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate"}
-sp-std = {branch = "polkadot-v0.9.10", default-features = false, git = "https://github.com/paritytech/substrate"}
+codec = {default-features = false, features = ["derive"], package = "parity-scale-codec", version = "2.3.1"}
```
Comment (weichweich) [inline pallets/ctype/Cargo.toml:31]: W: move `codec` and `scale-info` up and put it directly under [dependencies]. I would also group all optional dependencies together.

File: pallets/ctype/Cargo.toml [hunk 2]
```diff
@@ -14,36 +14,38 @@ substrate-wasm-builder-runner = {version = "3.0.0"}
...
+sp-keystore = {branch = "polkadot-v0.9.11", default-features = false, git = "https://github.com/paritytech/substrate", optional = true}
+sp-runtime = {branch = "polkadot-v0.9.11", default-features = false, git = "https://github.com/paritytech/substrate"}
+sp-std = {branch = "polkadot-v0.9.11", default-features = false, git = "https://github.com/paritytech/substrate"}
 
 [features]
 default = ["std"]
 mock = [
   "kilt-primitives",
   "pallet-balances",
+  # "scale-info",
```
Comment (weichweich) [inline pallets/ctype/Cargo.toml:48]: S: remove if it's not needed

File: pallets/ctype/src/mock.rs [hunk 3]
```diff
@@ -16,6 +16,8 @@
 
 // If you feel like getting in touch with us, you can do so at info@botlabs.org
 
+use crate as ctype;
+use crate::*;
```
Comment (weichweich) [inline pallets/ctype/src/mock.rs:20]: S: I would keep them in a separate group. In my mind imports are sorted like that:

1. external crates
2. crate from the same git
3. same crate
4. re-exports
5. mod

Also this is out of scope. Had a discussion with Antonio about that.

File: pallets/delegation/src/delegation_hierarchy.rs [hunk 4]
```diff
@@ -60,7 +61,9 @@ impl Default for Permissions {
 /// For quicker lookups of the hierarchy details, all nodes maintain a direct
 /// link to the hierarchy root node. Furthermore, all nodes have a parent except
 /// the root nodes, which point to themselves for the hierarchy root node link.
-#[derive(Clone, Encode, Decode, PartialEq)]
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+
```
Comment (weichweich) [inline pallets/delegation/src/delegation_hierarchy.rs:66]: S: why a new line here? Feels like it's not the rust style?

File: pallets/delegation/src/delegation_hierarchy.rs [hunk 5]
```diff
@@ -127,7 +130,9 @@ impl<T: Config> DelegationNode<T> {
 }
 
 /// Delegation information attached to delegation nodes.
-#[derive(Clone, Debug, Encode, Decode, PartialEq)]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+
```
Comment (weichweich) [inline pallets/delegation/src/delegation_hierarchy.rs:135]: S: why a new line here? Feels like it's not the rust style?

File: pallets/delegation/src/delegation_hierarchy.rs [hunk 6]
```diff
@@ -153,7 +158,9 @@ impl<T: Config> DelegationDetails<T> {
 }
 
 /// The details associated with a delegation hierarchy.
-#[derive(Clone, Debug, Encode, Decode, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Debug, Encode, Decode, Eq, PartialEq, Ord, PartialOrd, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+
```
Comment (weichweich) [inline pallets/delegation/src/delegation_hierarchy.rs:163]: S: why a new line here? Feels like it's not the rust style?

File: pallets/did/src/deprecated/v2.rs [hunk 7]
```diff
@@ -21,47 +21,50 @@ use codec::{Decode, Encode};
...
 pub(crate) struct ServiceEndpoints {
 	pub content_hash: Hash,
 	pub urls: Vec<Url>,
 	pub content_type: ContentType,
 }
 
-#[derive(Clone, Decode, Encode, PartialEq)]
+#[derive(Clone, Decode, Encode, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+
```
Comment (weichweich) [inline pallets/did/src/deprecated/v2.rs:67]: S: why a new line here? Feels like it's not the rust style?

File: pallets/did/src/did_details.rs [hunk 8]
```diff
@@ -220,7 +221,9 @@ impl DidVerifiableIdentifier for kilt_primitives::DidIdentifier {
 ///
 /// It is currently used to keep track of all the past and current
 /// attestation keys a DID might control.
-#[derive(Clone, Debug, Decode, Encode, PartialEq, Ord, PartialOrd, Eq)]
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Ord, PartialOrd, Eq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+
```
Comment (weichweich) [inline pallets/did/src/did_details.rs:226]: S: why a new line here? Feels like it's not the rust style?

File: pallets/kilt-launch/src/lib.rs [hunk 9]
```diff
@@ -678,38 +679,44 @@ pub mod pallet {
...
-				);
-				VestingInfo {
+			ensure!(
+				Vesting::<T>::decode_len(&target).unwrap_or_default() <= 1,
+				Error::<T>::MultipleVestingSchemes
+			);
+
+			let vesting =
+				if let Some(target_vesting) = Vesting::<T>::take(&target).unwrap_or_default().into_inner().first() {
+					// Should never throw because all source accounts start vesting in genesis block
```
Comment (weichweich) [inline pallets/kilt-launch/src/lib.rs:698]:
```suggestion
					// Should never return an error because all source accounts start vesting in genesis block
```

File: primitives/src/constants.rs [hunk 10]
```diff
@@ -119,7 +122,7 @@ pub mod staking {
 	pub const MIN_COLLATORS: u32 = 16;
 
 	#[cfg(feature = "fast-gov")]
-	pub const MAX_CANDIDATES: u32 = 16;
+	pub const MAX_CANDIDATES: u32 = 17;
```
Comment (weichweich) [inline primitives/src/constants.rs:125]: This is only active for `fast-gov`. I think this can remain 16?

File: pallets/ctype/Cargo.toml (hunk 1 above)
Comment (wischli) [reply pallets/ctype/Cargo.toml:31]: I meant to do that. Good catch!

File: pallets/did/src/did_details.rs [hunk 11]
```diff
@@ -220,7 +221,9 @@ impl DidVerifiableIdentifier for kilt_primitives::DidIdentifier {
 ///
 /// It is currently used to keep track of all the past and current
 /// attestation keys a DID might control.
-#[derive(Clone, Debug, Decode, Encode, PartialEq, Ord, PartialOrd, Eq)]
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Ord, PartialOrd, Eq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
```
[the remark this answers is not in the document]
Comment (JoshOrndorff) [reply pallets/did/src/did_details.rs:225]: Thank you for this helpful comment https://gist.github.com/ascjones/0d81a4c44e84cacd9f714cd34a6de823#gistcomment-3931135

I noticed you described it as `structs which lazily use the pallet's Config as generic`. Do you mean that the less lazy fix would be
```rust
pub struct DidPublicKeyDetails<DidPublicKey> {
	/// A public key the DID controls.
	pub key: DidPublicKey,
}
```

[the remark this answers is not in the document]
Comment (wischli) [reply pallets/did/src/did_details.rs:225]: Not sure the _lazy_ is the correct technical term but it meant to reflect that instead of specifying the different Generics, the Config trait is provider instead.
